Add GitHub actions build, fix file selector

linuxdaemon opened the pull request:
This PR adds a GitHub actions workflow to build the windows, mac, and linux binaries all at once.

This also includes:
- Fixing the file filters in browse file dialogs
- Pinning the dependency versions more strictly
- Adding a (temporary/placeholder) icon to satisfy macOS app package requirements
- Several other small changes

This is step 1 in moving to a fully automated build and deploy setup, this will build the binaries for each commit but only the repo admin has access to build artifacts by default. Next will be a workflow for releasing a new version which will create a github release and attach the binaries from the latest commit under that release. After that will just be a workflow to publish to PyPi

linuxdaemon committed:
Add github actions workflow to build binary

linuxdaemon committed:
Switch to nuitka-project annotations

- Enable macos app bundle
- Disable console
- Go back a version for the action

linuxdaemon committed:
Require pyside2 on non-windows systems

linuxdaemon committed:
Set better version constraints for deps

linuxdaemon committed:
Copy entire app bundle

linuxdaemon committed:
Test app icon

Add required imageio dep

Include extra macos build files

linuxdaemon committed:
Tar artifacts to preserve exec bit

linuxdaemon committed:
Add pkg building for macos

linuxdaemon committed:
Add version info for build, add pyproject.toml

linuxdaemon committed:
Update to nuitka factory to test fixes

Issue fixed: https://github.com/Nuitka/Nuitka/issues/2722

linuxdaemon committed:
Fix file selectors on mac/linux

julbean reviewed (commented):
[This](https://github.com/julbean/describealign/pull/2/commits/f2549326d06b1308b238b9b14ec87b56bafe2992) breaks it on Windows, but that can be fixed with a platform check (like on line 74)

julbean commented on pyproject.toml:
```diff
@@ -0,0 +1,37 @@
+  "scipy >= 1.10.1",
+  "pytsmod >= 0.3.7",
+  "PySimpleGUIWx==0.17.2 ; platform_system == 'Windows'",
+  "PySimpleGUIQt>=0.35.0 ; platform_system != 'Windows'",
```
We'll need to change this to "==" to keep it from using the new paid version

linuxdaemon replied:
Ah yes, I meant to switch this to read the requirements.txt so we only have one source of truth, for simplicity

linuxdaemon replied:
Fixed

julbean commented on requirements.txt:
```diff
@@ -5,5 +5,5 @@ numpy>=1.21.4
 pytsmod>=0.3.7
-PySimpleGUIWx==0.17.2; sys_platform == 'windows'
-PySimpleGUIQt>=0.35.0; sys_platform != 'windows'
+PySimpleGUIWx==0.17.2; platform_system == 'Windows'
```
This change seems unnecessary?  Also, the old version is consistent with the example in this documentation:
https://pip.pypa.io/en/stable/reference/requirement-specifiers/

linuxdaemon replied:
sys_platform would need to match 'win32', not 'Windows', I switched it to be consistent with the platform check in the code which uses platform.system() instead of sys.platform()

linuxdaemon replied:
I am fine switching this back to sys_platform, but we should keep platform checks consistent between dependencies and the code, up to you.

julbean replied:
So long as platform_system works, I think it's more clear so I'm fine with changing it.

julbean commented on describealign.py:
```diff
@@ -1,3 +1,19 @@
+# nuitka-project-if: {OS} != "Windows":
```
Not a big fan of this being above the "what does the script do" header, can it be moved below that?

linuxdaemon replied:
Yep, should work

linuxdaemon replied:
Done

julbean commented on describealign.py:
```diff
@@ -182,7 +198,7 @@ def tokenize_audio_dither(media_arr, slow_timings):
   fast_rate = 1. / (1 + 1. / (DITHER_PERIOD_STEPS - 2 + (np.sqrt(5) + 1) / 2.))
   fast_spec, fast_timings = tokenize_audio(media_arr, fast_rate)
-  
+
```
Don't mind the trailing white spaces being deleted, but I kind of like the empty lines having matching tabbing as it's less fiddly when editing.  Is there a reason to get rid of it?

linuxdaemon replied:
It's generally considered bad form as it adds extra characters to the file that have little use, as most editors will auto-indent when you add a new line in a code block.

julbean replied:
Found this discussion in which the top answer says PEP8 doesn't specify and both are legal, while the close second answer says it should be added as it's important for readability when display whitespace is enabled, especially with two-space indentation like we have here.
https://stackoverflow.com/questions/2727988/python-indentation-in-empty-lines

linuxdaemon replied:
This change was automatically implemented by VSCode, and reading through the code it seems to already be inconsistent. Is this a linter check we want to implement?

linuxdaemon replied:
That SO answer also mentions that the PEP8 doesn't take a stance on it, but under [this section](https://peps.python.org/pep-0008/#other-recommendations) it does specifically mention removing all trailing whitespace, not just whitespace after statements.

julbean replied:
Trailing whitespace is whitespace after statements.  I'd prefer to keep the non-trailing whitespace, but if it would bother you I also don't really mind that much either way.  Your call here.

linuxdaemon replied:
For now I have reverted these changes and added a vscode config file to disable removing trailing whitespace. As for the future, I think it would be beneficial to define a lint/formatter config either way, to enforce consistency.

As for removing the trailing whitespace on blank lines, `git diff --check` will yell at you about it by default, and most editors will automatically trim ALL trailing whitespace (including blank lines). Git specifically refers to both lines containing only whitespace and lines containing code followed by whitespace as 'lines with trailing whitespace', and thus default checks in the diff will alert on both. I think for this reason, we should migrate to stripping all trailing whitespace, to be consistent with the defaults in VSCode, Git, PyCharm, the `black` formatter, `ruff`, etc.

julbean commented on describealign.py:
```diff
@@ -1126,10 +1126,10 @@ def main_gui():
 
-  all_audio_file_types = [('All Audio File Types', '*.' + ';*.'.join(AUDIO_EXTENSIONS)),]
-  all_video_file_types = [('All Video File Types', '*.' + ';*.'.join(VIDEO_EXTENSIONS)),]
+  all_audio_file_types = [('All Audio File Types', '*.' + ' *.'.join(AUDIO_EXTENSIONS)),]
```
This breaks it on Windows, but that can be fixed with a platform check (like on line 74).

linuxdaemon replied:
Is this not handled by `file_fix` on 1139?

julbean replied:
That just deals with a bug in this bit of PySimpleGUIWx's code:
```
def convert_tkinter_filetypes_to_wx(filetypes):
    wx_filetypes = ''
    for item in filetypes:
        filetype = item[0] + ' (' + item[1] + ')|'+ item[1]
        wx_filetypes += filetype
    return wx_filetypes
```
The bug is it doesn't add the separator '|' between filetypes when it appends them to wx_filetypes, so file_fix just prepends it to every filetype after the first.
We can just do another if IS_RUNNING_WINDOWS, else here though and it should be fine.  Also I'm a bit surprised you were seeing a problem with the semicolon implementation on Mac/Linux, as I thought I'd tested it on there with the old version of PySimpleGuiQt.  Does it only need to be spaces for the new version or something?

linuxdaemon replied:
It worked fine on mac but on linux (ubuntu, opensuse, and a few other distros), it has to be space. space worked for mac as well, and [the QT docs](https://doc.qt.io/qt-6/qfiledialog.html#setNameFilters) only use spaces. I'll test windows and see about implementing that fix

linuxdaemon replied:
Fixed in 5410fb5

linuxdaemon committed:
Move nuitka-project comments below docs

linuxdaemon committed:
Move to reading deps from requirements.txt in pyproject.toml

linuxdaemon committed:
Set the filetype seperator on windows back to ';'

linuxdaemon committed:
Switch back to main nuitka version, includes action fixes now

julbean commented on describealign.py:
```diff
@@ -1128,10 +1128,11 @@ def main_gui():
 
-  all_audio_file_types = [('All Audio File Types', '*.' + ' *.'.join(AUDIO_EXTENSIONS)),]
-  all_video_file_types = [('All Video File Types', '*.' + ' *.'.join(VIDEO_EXTENSIONS)),]
+  filetype_sep = ';' if IS_RUNNING_WINDOWS else ' '
```
Nice use of a ternary here!  Very clean

linuxdaemon committed:
Preserve original indented blank lines

Also adds a vscode config to prevent the editor from automatically removing the indents

julbean commented:
Tested out the builds and none of them seem to run on my systems:
Windows 10: does/returns nothing and exits, no output when run in cmd prompt
MacOS 10.15.7 VM: "application requires macOS 11.7 or later"
Ubuntu 22.04.3 VM: crashes when trying to load the GUI:
```Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, webgl, xcb.
```

linuxdaemon commented:
Is it only the binaries or does running it from source error the same? @julbean

linuxdaemon commented:
Windows seems to detect anything built with --onefile as malware, so I think windows needs to stick to a standalone folder build. As for mac, I think just setting the github actions builder to an older version will work.

For ubuntu, are you running an ubuntu desktop basic install or ubuntu server with the ubuntu-desktop/DE packages installed by hand?

linuxdaemon commented:
@julbean macos <12 is officially unsupported by Apple, do we want to support such an old version?

julbean commented:
Given Mac runs software from any older version, it's basically a free lunch.  Compiling it on an older version like you suggested should fix it.  I'll test it on my install the first time here, but we don't need to officially support it or test it, so if it breaks down the line I don't think we'll worry about it.
For Ubuntu I'm just using a default ubuntu desktop installation.  If I installed the libraries needed to run the script, I'm pretty sure that would fix it (as I've seen people say installing pyqt5 or whatever fixes it), but that's not really the point of having a binary.

linuxdaemon commented:
The issue seems to be that `libxcb` is missing, I'm testing building with it installed and --include-qt-plugins set to include it. As for mac, I think we can't go too far back because of recent OS changes, but I can try 10.15 and see if it breaks or not

linuxdaemon commented:
> Given Mac runs software from any older version, it's basically a free lunch. Compiling it on an older version like you suggested should fix it. I'll test it on my install the first time here, but we don't need to officially support it or test it, so if it breaks down the line I don't think we'll worry about it. For Ubuntu I'm just using a default ubuntu desktop installation. If I installed the libraries needed to run the script, I'm pretty sure that would fix it (as I've seen people say installing pyqt5 or whatever fixes it), but that's not really the point of having a binary.

Ok so GitHub Actions only has runners for macOS 11, 12, 13, and 14. Looks like we can't build for 10.x macOS on github without a custom runner (read: $$).

julbean commented:
Aww, well that's okay I can always just compile a Mac 10.x version and upload it manually
